File: 2022/day07.ts
import { example, load, solve } from 'lib/advent';
import { iter } from 'lib/iter';

type Cmd = { cmd: string[]; output: string[][] };

function parse(input: string): Cmd[] {
  return input
    .split('$ ')
    .slice(1)
    .map((chunk) => {
      const [cmd, ...output] = chunk
        .split('\n')
        .map((line) => line.split(/\s+/));
      return { cmd, output };
    });
}

class Dir {
  children: Dir[] = [];
  files: { name: string; size: number }[] = [];

  constructor(public name: string, public parent?: Dir) {}

  get size(): number {
    return iter([...this.files, ...this.children])
      .pluck('size')
      .sum();
  }

  walk(): Dir[] {
    return [this, ...this.children.flatMap((x) => x.walk())];
  }
}

function tree(cmds: Cmd[]): Dir {
  const root = new Dir('/');
  let cwd = root;
  for (const { cmd, output } of cmds) {
    if (cmd[0] === 'cd') {
      const path = cmd[1];
      if (path === '/') cwd = root;
      else if (path === '..') cwd = cwd.parent;
      else cwd = cwd.children.find((d) => d.name === path);
    } else if (cmd[0] === 'ls') {
      const [dirs, files] = iter(output).partition((line) => line[0] === 'dir');
      cwd.children.push(...dirs.map((line) => new Dir(line[1], cwd)));
      cwd.files.push(
        ...files.map((line) => ({ size: Number(line[0]), name: line[1] }))
      );
    }
  }
  return root;
}

function sumOfSmallDirs(root: Dir): number {
  return iter(root.walk())
    .pluck('size')
    .filter((x) => x <= 100000)
    .sum();
}

function toDelete(root: Dir): number {
  const needed = 30_000_000 - (70_000_000 - root.size);
  return iter(root.walk())
    .pluck('size')
    .filter((x) => x >= needed)
    .min();
}

const exampleData = parse(load('ex').raw);
example.equal(95437, sumOfSmallDirs(tree(exampleData)));
example.equal(24933642, toDelete(tree(exampleData)));

const data = parse(load().raw);
export default solve(
  () => sumOfSmallDirs(tree(data)),
  () => toDelete(tree(data))
).expect(1989474, 1111607);

File: py/2022/day07.py
from aoc import main


def parse(s: str) -> dict[str, int]:
    tree = {"/": 0}
    pwd = ["/"]
    for line in s.splitlines():
        match line.split(" "):
            case "$", "cd", dirname if dirname == "..":
                pwd.pop()
            case "$", "cd", dirname if dirname != "/":
                pwd.append(f"{pwd[-1]}/{dirname}")
                tree[pwd[-1]] = 0
            case ("$", "ls") | ("dir", _):
                pass
            case size, _:
                for path in pwd:
                    tree[path] += int(size)
    return tree


def to_delete(tree: dict[str, int], fs_size=70000000, req_size=30000000) -> int:
    needed = req_size - (fs_size - tree["/"])
    return min(s for s in tree.values() if s >= needed)


if __name__ == "__main__":
    main(
        lambda s: sum(size for size in parse(s).values() if size < 100000),
        lambda s: to_delete(parse(s)),
    )
